File: markov.js
const { table } = require('console');
const util = require('util');

// An array maps indices to values. Inverting it means returning a map of values to indices.
const invertArray = arr => {
    const inverted = {};
    for (let i = 0; i < arr.length; i++) {
        inverted[arr[i]] = i;
    }
    return inverted;
}

const gcd = (a, b) => {
    if (b > a) {
        const tmp = a;
        a = b;
        b = tmp;
    }
    while (true) {
        if (b === 0) return a;
        a %= b;
        if (a === 0) return b;
        b %= a;
    }
}

const generateMarkovIntoTable = (str, order, table, chars, ngrams) => {
    for (let i = 0; i < str.length - order; i++) {
        const currentNgram = str.slice(i, i + order);
        ngrams.add(currentNgram);
        const nextChar = str[i + order];
        chars.add(nextChar);

        let tableEntry;
        if (Object.prototype.hasOwnProperty.call(table, currentNgram)) {
            tableEntry = table[currentNgram];
            if (Object.prototype.hasOwnProperty.call(tableEntry, nextChar)) {
                tableEntry[nextChar]++;
            } else {
                tableEntry[nextChar] = 1;
            }
        } else {
            tableEntry = {[nextChar]: 1};
            table[currentNgram] = tableEntry;
        }

        if (i === (str.length - order) - 1) {
            // make sure we get the *last* ngram too
            ngrams.add(str.slice(i + 1));
        }
    }
}

// Generate a Markov chain probability table from a string or array of strings
const generateMarkov = (strings, order = 2) => {
    const table = {};
    const chars = new Set();
    const ngrams = new Set();
    if (Array.isArray(strings)) {
        for (const str of strings) {
            generateMarkovIntoTable(str, order, table, chars, ngrams);
        }
    } else {
        generateMarkovIntoTable(strings, order, table, chars, ngrams);
    }

    const ngramsArr = Array.from(ngrams).sort();
    const ngramIndices = invertArray(ngramsArr);

    // Convert string-based table (slow) to numeric indices (fast)
    const markovArr = [];

    for (let i = 0; i < ngramsArr.length; i++) {
        const ngram = ngramsArr[i];
        const probs = table[ngram];
        const probsArr = [];
        let totalProbability = 0;
        for (const nextChar of Object.keys(probs)) {
            const nextNgram = ngram.slice(1) + nextChar;
            const probability = probs[nextChar];
            totalProbability += probability;
            probsArr.push({
                nextNgram: ngramIndices[nextNgram],
                nextChar,
                probability
            });
        }
        // sort by probability in descending order to minimize linear search steps
        probsArr.sort((a, b) => b.probability - a.probability);
        let sum = 0;
        for (let i = 0; i < probsArr.length; i++) {
            sum += probsArr[i].probability;
            probsArr[i].cumulativeProbability = sum;
            delete probsArr[i].probability;
        }
        let probGCD = probsArr[0].cumulativeProbability;
        if (probsArr.length > 1) {
            for (let i = 1; i < probsArr.length; i++) {
                probGCD = gcd(probGCD, probsArr[i].cumulativeProbability);
            }
        }
        if (probGCD > 1) {
            for (const entry of probsArr) {
                entry.cumulativeProbability /= probGCD;
            }
            totalProbability /= probGCD;
        }

        markovArr.push({choices: probsArr, totalProbability});
    }

    return {
        table,
        markovArr,
        order,
        ngrams: {values: ngramsArr, indices: ngramIndices}
    };
}

const weightedRand = (table) => {
    const tableSum = Object.values(table).reduce((prev, cur) => prev + cur, 0);
    const random = Math.floor(Math.random() * tableSum);
    let sum = 0;
    for (const key of Object.keys(table)) {
        sum += table[key];
        if (random < sum) return key;
    }
}

const generateFromTable = ({table, order}, len, start) => {
    let generated = start;

    while (generated.length < len) {
        const prev = generated.slice(generated.length - order, generated.length);
        if (!Object.prototype.hasOwnProperty.call(table, prev)) return generated;
        const nextOptions = table[prev];

        generated += weightedRand(nextOptions);
    }

    return generated;
}

const generateFromArray = ({markovArr: table, ngrams}, len, start) => {
    let generated = start;
    let ngram = ngrams.indices[start];

    while (generated.length < len) {
        const {choices, totalProbability} = table[ngram];
        const random = Math.floor(Math.random() * totalProbability);
        for (let i = 0; i < choices.length; i++) {
            const {nextChar, nextNgram, cumulativeProbability} = choices[i];
            if (random < cumulativeProbability) {
                ngram = nextNgram;
                generated += nextChar;
                break;
            }
        }
    }

    return generated;
}

const markovArrToC = ({markovArr: table, ngrams}, name) => {
    const choicesDefs = [];
    const ngramDefs = [];

    for (let i = 0; i < table.length; i++) {
        const listName = `${name}_ngram${i}_choices`;
        const {choices, totalProbability} = table[i];
        const choiceDefs = [];
        for (let i = 0; i < choices.length; i++) {
            const {nextChar, nextNgram, cumulativeProbability} = choices[i];
            choiceDefs.push(
`    {.next_ngram = ${nextNgram}, .cumulative_probability = ${cumulativeProbability}, .next_char = '${nextChar}'}`
);
        }

        choicesDefs.push(
`static struct uwu_markov_choice ${listName}[] = {
${choiceDefs.join(',\n')}
};`
        );

        ngramDefs.push(`    {.choices = ${listName}, .total_probability = ${totalProbability}}${i === table.length - 1 ? '' : ','} // ${ngrams.values[i]}`);
    }

    const code =
`${choicesDefs.join('\n')}

static uwu_markov_ngram ${name}_ngrams[] = {
${ngramDefs.join('\n')}
};`;

   return code;
}

const catgirlNonsense = `mraowmraowmewmraowmrrppurrrrmraownyanyamraowwwwwmrwmraowmreowmewmrowmraowmewmraownya
mrrrowmeowmrowrmrowpurrrmrowmeowmraowmewmrowrmewnyaaaamrowrnyaamewmeowmrowrmeownya
mrowrmeow
meowmrowrmrowmeow
mrowrnyaaaaa
mrrowmeowmrowrmrrowmeowmrowrnyaaamrowmewmeowmewpurrrrrmeowmrowrmeownya`;

const keysmash = `alksdhfl;ag;kdhfgjkfhgadskfagdfkajfdhgbklkafghkahgsdfka;dfglkfjhgajdfghkgahjfgafgfkjdhg;lskgjhjkhajhdgfjhkafgl;ajdfglkajdflg;hdkjafhgkgaurgjrahdfgbahiurghrgh;arhnguahraufjalfgnhjhaujeghfgadjog;aldhhjlahuegjfdbhgajkfghafkjgahiurg`;

const scrunklyBase = "the little tienpsy! so adornale and cutest tootsit! awww like and double tap now so it can the to live the cute! happy kdb! awww the scrunkly! scrunkly the when! the boinky spunge! crinkly doo! shronkle scrimblo! aww when the.. the limtle tootsie,, tienpsy widdle scrimblo boinkus! boinky spunge! crinkly doo! scrunkly,,, the widdle. the cutest adornale tienpsy tootsit,,, whem the kity,, n flunf and it,, i,. yay! lookit aw!! lookit the little crungle boinko! aw icamt,,, the kimty and cat sooo mipy. little meowmeow tienpsy and smol sproingle scrunkly"

const punctuationRegex = /[,.!]+ ?/g;

const scrunks = [];

const xoshiro128 = (() => {
    const s = [1697843356, 1544689657, 3646425737, 2133148247];

    const rotl = (x, k) => (x << k) | (x >> (32 - k));

    return () => {
        const x = s[0] + s[3];
        const result = (rotl(s[0] + s[3], 7) + s[0]) & 0xffffffff;

        const t = s[1] << 9;
        s[2] ^= s[0];
        s[3] ^= s[1];
        s[1] ^= s[2];
        s[0] ^= s[3];

        s[2] ^= t;

        s[3] = rotl(s[3], 11);

        return result >>> 0;
    }
})();

for (let i = 0; i < 100; i++) {
    scrunks.push(scrunklyBase.replace(punctuationRegex, match => {
        if (xoshiro128() % 3 === 0) {
            return ','.repeat((xoshiro128() % 3) + 2);
        }
        return ['!', '.'][xoshiro128() % 2].repeat((xoshiro128() % 4) + 1) + ' ';
    }))
}

const catgirlTable = generateMarkov(catgirlNonsense.split('\n'), 2)
const keysmashTable = generateMarkov(keysmash, 1);
const scrunklyTable = generateMarkov(scrunks, 2);

/*console.log(generateFromArray(catgirlTable, 100, 'ny'));

console.time('from table');
for (let i = 0; i < 1000; i++) {
    generateFromTable(catgirlTable, 1000, 'ny');
}
console.timeEnd('from table');

console.time('from array');
for (let i = 0; i < 1000; i++) {
    generateFromArray(catgirlTable, 1000, 'ny');
}
console.timeEnd('from array');*/

console.log(markovArrToC(catgirlTable, 'catnonsense'));
console.log(markovArrToC(keysmashTable, 'keysmash'));
console.log(markovArrToC(scrunklyTable, 'scrunkly'));

File: markov.py
from typing import Tuple, TypeVar, Generic, OrderedDict, Optional
from dataclasses import dataclass
from enum import Enum
import re

T = TypeVar('T')

@dataclass
class NgramChoice:
    next_ngram: int
    cumulative_probability: int

@dataclass
class Ngram(Generic[T]):
    choices: list[NgramChoice]
    total_probability: int
    chars: Tuple[T, ...]

class SpecialMarkov(Enum):
    SCRUNKLY_PUNCTUATION = -1

class MarkovTable(Generic[T]):
    def __init__(self, order: int, *strs: list[T]):
        self.entries: OrderedDict[Tuple[T, ...], dict[T, int]] = OrderedDict()
        self.order = order

        for str in strs:
            self.add_string(str)

    def add_string(self, str: list[T]):
        for i in range(0, len(str) - self.order):
            ngram = tuple(str[i:i + self.order])
            next_char = str[i + self.order]

            ngram_entry = self.entries.setdefault(ngram, {})
            if next_char in ngram_entry:
                ngram_entry[next_char] += 1
            else:
                ngram_entry[next_char] = 1

        # TODO: emit terminal state on last tuple

    def to_table(self) -> Tuple[list[Ngram[T]], dict[Tuple[T, ...], int]]:
        # sorted_entries = OrderedDict(sorted(self.entries.items(), key = lambda item: ''.join(item[0]))) # type: ignore
        sorted_entries = self.entries

        table: list[Ngram[T]] = []
        ngram_indices = {ngram: index for index, ngram in enumerate(sorted_entries.keys())}
        for index, (ngram, probabilities) in enumerate(sorted_entries.items()):

            total_probability = sum(probabilities.values())
            if total_probability > 255:
                raise OverflowError('Total probability overflows a uint8')
            # sort by probability in descending order to minimize linear search steps
            next_chars = sorted(probabilities.items(), key=lambda p: p[1], reverse=True)

            cumulative_probability = 0
            choices = []
            for char, probability in next_chars:
                next_ngram = ngram[1:] + (char,)
                if next_ngram not in ngram_indices:
                    raise ValueError(f'Found terminal ngram {next_ngram}')

                cumulative_probability += probability
                next_ngram_index = ngram_indices[next_ngram]
                if next_ngram_index > 255:
                    raise OverflowError('Ngram index overflows a uint8')
                if cumulative_probability > 255:
                    raise OverflowError('Probability overflows a uint8')
                choice = NgramChoice(next_ngram_index, cumulative_probability)
                choices.append(choice)

            table.append(Ngram(choices, total_probability, ngram))

        return table, ngram_indices

    def to_c(self, prefix: str, initial_ngram: Optional[Tuple[T, ...]] = None, specials: Optional[str] = None) -> str:
        markov_table, ngram_indices = self.to_table()

        choice_defs = ["static const uwu_markov_choice {prefix}_choices[] = {{".format(prefix=prefix)]
        ngram_defs = ["static const uwu_markov_ngram {prefix}_ngrams[] = {{".format(prefix=prefix)]

        table_def = """static const uwu_markov_table {prefix}_table = {{
    .specials = {specials},
    .initial_ngram={initial_ngram},
    .num_ngrams={num_ngrams},
    .choices={prefix}_choices,
    .ngrams={prefix}_ngrams
}};""".format(
            prefix=prefix,
            specials=specials if specials is not None else 'NULL',
            initial_ngram=ngram_indices[initial_ngram] if initial_ngram is not None else -1,
            num_ngrams=len(markov_table)
    )

        choice_table_index = 0
        for ngram_index, ngram in enumerate(markov_table):
            choice_defs.append('    // ngram {index} {chars}'.format(index=ngram_index, chars=ngram.chars))
            for choice_index, choice in enumerate(ngram.choices):
                choice_defs.append('    {{.next_ngram = {next_ngram}, .cumulative_probability = {cumulative_probability}}},'.format(
                    next_ngram=choice.next_ngram,
                    cumulative_probability=choice.cumulative_probability,
                    comma='' if choice_index == len(ngram.choices) - 1 and ngram_index == len(markov_table) - 1 else ','
                ))

            last_char = ngram.chars[-1]

            ngram_defs.append('    {{.choices = {index}, .total_probability = {total_probability}, .character = {character}}}, // "{chars}"'.format(
                prefix=prefix,
                index=choice_table_index,
                total_probability=ngram.total_probability,
                character=last_char.value if isinstance(last_char, SpecialMarkov) else f"'{last_char}'",
                chars=ngram.chars
            ))

            choice_table_index += len(ngram.choices)

        ngram_defs.append('};')
        choice_defs.append('};')

        return '{choices}\n\n{ngrams}\n\n{table_def}'.format(choices='\n'.join(choice_defs), ngrams='\n'.join(ngram_defs), table_def=table_def)


catgirl_nonsense = [list(s) for s in """mraowmraowmewmraowmrrppurrrrmraownyanyamraowwwwwmrwmraowmreowmewmrowmraowmewmraownya
mrrrowmeowmrowrmrowpurrrmrowmeowmraowmewmrowrmewnyaaaamrowrnyaamewmeowmrowrmeownya
mrowrmeow
meowmrowrmrowmeow
mrowrnyaaaaa
mrrowmeowmrowrmrrowmeowmrowrnyaaamrowmewmeowmewpurrrrrmeowmrowrmeownya""".split()]

keysmash = list('alksdhfl;ag;kdhfgjkfhgadskfagdfkajfdhgbklkafghkahgsdfka;dfglkfjhgajdfghkgahjfgafgfkjdhg;lskgjhjkhajhdgfjhkafgl;ajdfglkajdflg;hdkjafhgkgaurgjrahdfgbahiurghrgh;arhnguahraufjalfgnhjhaujeghfgadjog;aldhhjlahuegjfdbhgajkfghafkjgahiurg')

scrunkly_base = 'the little tienpsy! so adornale and cutest tootsit! awww like and double tap now so it can the to live the cute! happy kdb! awww the scrunkly! scrunkly the when! the boinky spunge! crinkly doo! shronkle scrimblo! aww when the.. the limtle tootsie,, tienpsy widdle scrimblo boinkus! boinky spunge! crinkly doo! scrunkly,,, the widdle. the cutest adornale tienpsy tootsit,,, whem the kity,, n flunf and it,, i,. yay! lookit aw!! lookit the little crungle boinko! aw icamt,,, the kimty and cat sooo mipy. little meowmeow tienpsy and smol sproingle scrunkly,, the liddol blibbly spoinky bintle anmd floof,,, flumfy! sooo springly poinkle n liddol mewmew,,,babey,,tha bleeb,,,blinpus,,, the springle!! widdol liddle kity sooo smol,,, teeny tiney meow kity'

punctuation = re.compile('[,.!]+\\s*')

scrunkly_chars = []
prev_match_start = 0
for match in punctuation.finditer(scrunkly_base):
    scrunkly_chars += scrunkly_base[prev_match_start:match.start()]
    prev_match_start = match.end()
    scrunkly_chars.append(SpecialMarkov.SCRUNKLY_PUNCTUATION)

if prev_match_start < len(scrunkly_base):
    scrunkly_chars += scrunkly_base[prev_match_start:]


catgirl_table = MarkovTable(2, *catgirl_nonsense)
keysmash_table = MarkovTable(1, keysmash)
scrunkly_table = MarkovTable(2, scrunkly_chars)

with open('uwurandom_markov_data.h', 'w', encoding='utf-8') as f:
    f.write("""#ifndef _UWURANDOM_MARKOV_DATA_H
#define _UWURANDOM_MARKOV_DATA_H

#include "uwurandom_types.h"
#include "uwurandom_markov_special.h"

""")
    f.write(catgirl_table.to_c('catnonsense', ('m', 'r')))
    f.write('\n\n')
    f.write(keysmash_table.to_c('keysmash'))
    f.write('\n\n')
    f.write(scrunkly_table.to_c('scrunkly', ('a', 'w'), 'scrunkly_specials'))
    f.write("\n\n#endif\n")
